crates/bevy_ecs/src/component.rs

```diff
@@ -952,6 +1000,283 @@ impl Components {
         self.components.get_mut(id.0).map(|info| &mut info.hooks)
     }
 
+    #[inline]
+    pub(crate) fn get_required_components_mut(
+        &mut self,
+        id: ComponentId,
+    ) -> Option<&mut RequiredComponents> {
+        self.components
+            .get_mut(id.0)
+            .map(|info| &mut info.required_components)
+    }
+
+    /// Recursively registers the given component `R` and its [required components] as required by `T`.
+    ///
+    /// When `T` is added to an entity, `R` will also be added if it was not already provided.
+    /// The given `constructor` will be used for the creation of `R`.
+    ///
+    /// [required components]: Component#required-components
+    ///
+    /// # Panics
+    ///
+    /// Panics if the `required` component is already a directly required component for the `requiree`.
+    ///
+    /// Indirect requirements through other components are allowed. In those cases, any existing requirements
+    /// will only be overwritten if the new requirement is more specific.
+    pub(crate) fn register_required_components_recursive<R: Component>(
+        &mut self,
+        required: ComponentId,
+        requiree: ComponentId,
+        constructor: fn() -> R,
+    ) -> Result<(), RequiredComponentsError> {
+        // SAFETY: We just created the components.
+        unsafe {
+            self.register_required_components_recursive_unchecked(required, requiree, constructor)
+        }
+    }
```

Comment:
Whoops, missed this when refactoring some of the logic to move the component initialization out of this method. I just removed this method now, and use the unsafe version directly (since this should've been technically unsafe anyway).